File: Dynamic Programming/acmicpc.net/17404/Main.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

class Main {
    static final String inputFilePath = "./Dynamic Programming/acmicpc.net/17404/sample.txt";
    // static final int mod = 1000000000;
    // static final String inputFilePath = "./# format/sample.txt";

    public static void main(String[] args) throws Exception {
        System.setIn(new FileInputStream(inputFilePath));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int N = Integer.parseInt(br.readLine());
        int a[][] = new int[N + 1][3];
        int d[][] = new int[N + 1][3];
        for (int i = 1; i <= N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            a[i][0] = Integer.parseInt(st.nextToken());
            a[i][1] = Integer.parseInt(st.nextToken());
            a[i][2] = Integer.parseInt(st.nextToken());
        }
        int answer = Integer.MAX_VALUE;
        for (int j = 0; j <= 2; j++) {
            if (j == 0) {
                d[1][0] = a[1][j];
                d[1][1] = 99999;
                d[1][2] = 99999;
            } else if (j == 1) {
                d[1][0] = 99999;
                d[1][1] = a[1][j];
                d[1][2] = 99999;
            } else if (j == 2) {
                d[1][0] = 99999;
                d[1][1] = 99999;
                d[1][2] = a[1][j];
            }

            for (int i = 2; i <= N; i++) {
                d[i][0] = Math.min(d[i - 1][1], d[i - 1][2]) + a[i][0];
                d[i][1] = Math.min(d[i - 1][0], d[i - 1][2]) + a[i][1];
                d[i][2] = Math.min(d[i - 1][0], d[i - 1][1]) + a[i][2];
            }
            int minValue = 0;
            if (j == 0) {
                minValue = Math.min(d[N][1], d[N][2]);
            } else if (j == 1) {
                minValue = Math.min(d[N][0], d[N][2]);
            } else if (j == 2) {
                minValue = Math.min(d[N][0], d[N][1]);
            }

            if (answer > minValue) {
                answer = minValue;
            }
        }
        System.out.println(answer);
    }
}

File: Dynamic Programming/acmicpc.net/17404/Main.js
const inputFilePath = "/sample.txt";
let input = require("fs")
  .readFileSync(__dirname + inputFilePath)
  // .readFileSync('/dev/stdin')
  .toString()
  .trim()
  .split("\n");

let N = input.shift();
let a = [];
let d = [];
for (let i = 0; i < N; i++) {
  d.push([]);
  a[i] = input[i].split(' ').map(item => {
    return parseInt(item);
  });
}
a.unshift([]);
d.unshift([]);
let answer = 999999;
for (let i = 0; i <= 2; i++) {
  for (let k = 0; k <= 2; k++) {
    if (i === k) {
      d[1][k] = a[1][i];
    } else {
      d[1][k] = 99999999;
    }
  }
  let minValue = 9999999;
  for (let j = 2; j <= N; j++) {
    d[j][0] = Math.min(d[j - 1][1], d[j - 1][2]) + a[j][0];
    d[j][1] = Math.min(d[j - 1][0], d[j - 1][2]) + a[j][1];
    d[j][2] = Math.min(d[j - 1][0], d[j - 1][1]) + a[j][2];
  }
  for (let k = 0; k <= 2; k++) {
    if (i === k) continue;
    minValue = Math.min(minValue, d[N][k]);
  }
  if (answer > minValue) {
    answer = minValue;
  }
}
console.log(answer);
